src/runtime/virtcontainers/qemu.go

```diff
@@ -1020,47 +1053,199 @@ func (q *qemu) togglePauseSandbox(pause bool) error {
+		EventCh: events,
+	}
 
 	// Auto-closed by QMPStart().
 	disconnectCh := make(chan struct{})
 
-	qmp, _, err := govmmQemu.QMPStart(q.qmpMonitorCh.ctx, q.qmpMonitorCh.path, cfg, disconnectCh)
+	qmp, _, err := govmmQemu.QMPStart(qmpChan.ctx, qmpChan.path, cfg, disconnectCh)
 	if err != nil {
 		q.Logger().WithError(err).Error("Failed to connect to QEMU instance")
 		return err
 	}
 
-	err = qmp.ExecuteQMPCapabilities(q.qmpMonitorCh.ctx)
+	err = qmp.ExecuteQMPCapabilities(qmpChan.ctx)
 	if err != nil {
 		qmp.Shutdown()
 		q.Logger().WithError(err).Error(qmpCapErrMsg)
 		return err
 	}
-	q.qmpMonitorCh.qmp = qmp
-	q.qmpMonitorCh.disconn = disconnectCh
+	qmpChan.qmp = qmp
+	qmpChan.disconn = disconnectCh
+
+	return nil
+}
+
+func (q *qemu) loopQMPEvent(event chan govmmQemu.QMPEvent) {
+	for {
+		select {
+		case e, open := <-event:
+			if !open {
+				q.Logger().Infof("QMP event channel closed")
+				return
+			}
+			q.Logger().Debugf("got new QMP event %+v", e)
+			if e.Name == "GUEST_PANICKED" {
+				go q.handleGuestPanic()
+			}
+		}
+	}
+}
+
+func (q *qemu) handleGuestPanic() {
+	if q.config.GuestMemoryDumpPath != "" {
+		q.dumpGuestMemory(q.config.GuestMemoryDumpPath)
+	}
+	// TODO: how to notify the upper level sandbox to handle the error
+	// to do a fast fail(shutdown or others).
+	// tracked by https://github.com/kata-containers/kata-containers/issues/1026
+}
+
+func (q *qemu) canDumpGuestMemory(dumpSavePath string) error {
+	fs := unix.Statfs_t{}
+	if err := unix.Statfs(dumpSavePath, &fs); err != nil {
+		q.Logger().WithError(err).WithField("dumpSavePath", dumpSavePath).Error("failed to call Statfs")
+		return nil
+	}
+	availSpaceInBytes := fs.Bavail * uint64(fs.Bsize)
+	q.Logger().WithFields(
+		logrus.Fields{
+			"dumpSavePath":      dumpSavePath,
+			"availSpaceInBytes": availSpaceInBytes,
+		}).Info("get avail space")
+
+	// get guest memory size
+	guestMemorySizeInBytes := (uint64(q.config.MemorySize) + uint64(q.state.HotpluggedMemory)) << utils.MibToBytesShift
+	q.Logger().WithField("guestMemorySizeInBytes", guestMemorySizeInBytes).Info("get guest memory size")
+
+	// default we want ensure there are at least double of VM memory size free spaces available,
+	// this may complete one dump operation for one sandbox
+	exceptMemorySize := guestMemorySizeInBytes * 2
+	if availSpaceInBytes >= exceptMemorySize {
+		return nil
+	} else {
+		return fmt.Errorf("there are not enough free space to store memory dump file. Except %d bytes, but only %d bytes available", exceptMemorySize, availSpaceInBytes)
+	}
+}
+
+func (q *qemu) dumpGuestMemory(dumpSavePath string) error {
+	q.memoryDumpFlag.Lock()
+	defer q.memoryDumpFlag.Unlock()
+
+	q.Logger().WithField("dumpSavePath", dumpSavePath).Info("try to dump guest memory")
+
+	dumpSavePath = filepath.Join(dumpSavePath, q.id)
+	dumpStatePath := filepath.Join(dumpSavePath, "state")
+	if fi, err := os.Stat(dumpStatePath); err != nil {
+		if os.IsNotExist(err) {
+			os.MkdirAll(dumpStatePath, 0644)
```

Comment:
- Missing check on return code of `MkdirAll()`.
- Perms should be `0600` I think.